minecraft/protocol/packet/compression.go

```diff
@@ -106,20 +108,43 @@ func (flateCompression) Compress(decompressed []byte) ([]byte, error) {
 
 // Decompress ...
 func (flateCompression) Decompress(compressed []byte, limit int) ([]byte, error) {
-	buf := bytes.NewReader(compressed)
-	c := flateDecompressPool.Get().(io.ReadCloser)
-	defer flateDecompressPool.Put(c)
+	r := flateDecompressPool.Get().(io.ReadCloser)
+	defer func() {
+		_ = r.Close()
+		flateDecompressPool.Put(r)
+	}()
 
-	if err := c.(flate.Resetter).Reset(buf, nil); err != nil {
+	if err := r.(flate.Resetter).Reset(bytes.NewReader(compressed), nil); err != nil {
 		return nil, fmt.Errorf("reset flate: %w", err)
 	}
-	_ = c.Close()
 
-	// Guess an uncompressed size of 2*len(compressed).
-	decompressed := bytes.NewBuffer(make([]byte, 0, len(compressed)*2))
-	if _, err := io.Copy(decompressed, io.LimitReader(c, int64(limit))); err != nil {
+	var decompressed bytes.Buffer
+	// If the compressed data is less than half the limit, we can safely assume l*2, otherwise cap at limit.
+	l := len(compressed)
+	capHint := limit
+	if l <= limit/2 {
+		capHint = l * 2
+	}
+	decompressed.Grow(capHint)
+
+	// Handle no limit
+	if limit == math.MaxInt {
+		if _, err := io.Copy(&decompressed, r); err != nil {
+			return nil, fmt.Errorf("decompress flate: %w", err)
+		}
+		return decompressed.Bytes(), nil
+	}
+
+	// Read limit+1 bytes to detect overflow without CopyN truncating the result.
+	toRead := int64(limit) + 1
+	n, err := io.CopyN(&decompressed, r, toRead)
+	if err != nil && err != io.EOF {
+		// CopyN returns an EOF error if the stream is shorter than the limit, we can treat that as a success.
 		return nil, fmt.Errorf("decompress flate: %w", err)
 	}
+	if n > int64(limit) {
+		return nil, fmt.Errorf("decompress flate: size %d exceeds limit %d", n, limit)
+	}
```

Comment:
Maybe a slightly easier to understand implementation, slightly closer to what it was before, might be this?
```go
lr := io.LimitReader(c, int64(limit) + 1)
if _, err := io.Copy(decompressed, lr); err != nil {
    return nil, fmt.Errorf("decompress flate: %w", err)
} else if lr.(*io.LimitedReader).N <= 0 {
    return nil, fmt.Errorf("decompress flate: size %d exceeds limit %d", n, limit)
}
```